# Movies Tonight Part 3: SQL DML

In [2]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

#### we did this in the terminal 

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [3]:
%%sql
INSERT INTO MOVIES (MTitle,Rating)
SELECT DISTINCT MTitle, Rating
FROM DATASET;

 * sqlite:///Movies.db
23 rows affected.


[]

In [4]:
%%sql
INSERT INTO ARTISTS (Name)
SELECT DISTINCT Cname
FROM DATASET;

 * sqlite:///Movies.db
153 rows affected.


[]

In [5]:
%%sql
INSERT INTO THEATERS(Tname,Phone,Location)
SELECT DISTINCT Tname, Phone, Location
FROM DATASET;

 * sqlite:///Movies.db
9 rows affected.


[]

In [6]:
%%sql
INSERT INTO CREDITS (MID,AID,CCode)
SELECT DISTINCT MID, AID, CCODE 
FROM DATASET 
    JOIN MOVIES ON (MOVIES.MTitle=DATASET.MTitle)
    JOIN ARTISTS ON (ARTISTS.Name=DATASET.CName)
    ;

 * sqlite:///Movies.db
162 rows affected.


[]

In [7]:
%%sql
INSERT INTO SHOWS(MID,TID,STime)
SELECT DISTINCT MID, TID, ShowTime 
FROM DATASET 
    JOIN MOVIES ON (MOVIES.MTitle=DATASET.MTitle)
    JOIN THEATERS ON (THEATERS.TName=DATASET.TName)
    ;

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [14]:
%%sql
SELECT *
    FROM CREDITS;

 * sqlite:///Movies.db
Done.


[(1, 'A', 1, 1),
 (2, 'A', 2, 1),
 (3, 'A', 3, 1),
 (4, 'A', 4, 1),
 (5, 'A', 5, 1),
 (6, 'A', 6, 1),
 (7, 'A', 7, 1),
 (8, 'A', 8, 1),
 (9, 'D', 9, 1),
 (10, 'A', 10, 2),
 (11, 'A', 11, 2),
 (12, 'A', 12, 2),
 (13, 'A', 13, 2),
 (14, 'A', 14, 2),
 (15, 'A', 15, 2),
 (16, 'A', 16, 2),
 (17, 'A', 17, 2),
 (18, 'D', 18, 2),
 (19, 'A', 19, 3),
 (20, 'A', 20, 3),
 (21, 'A', 21, 3),
 (22, 'A', 22, 3),
 (23, 'A', 23, 3),
 (24, 'A', 24, 3),
 (25, 'A', 25, 3),
 (26, 'A', 26, 3),
 (27, 'A', 27, 3),
 (28, 'A', 28, 3),
 (29, 'A', 29, 3),
 (30, 'D', 30, 3),
 (31, 'A', 31, 4),
 (32, 'A', 32, 4),
 (33, 'A', 33, 4),
 (34, 'A', 34, 4),
 (35, 'A', 35, 4),
 (36, 'D', 35, 4),
 (37, 'A', 36, 5),
 (38, 'A', 37, 5),
 (39, 'A', 38, 5),
 (40, 'A', 39, 5),
 (41, 'A', 40, 5),
 (42, 'A', 41, 5),
 (43, 'A', 42, 5),
 (44, 'D', 43, 5),
 (45, 'A', 44, 6),
 (46, 'A', 45, 6),
 (47, 'A', 46, 6),
 (48, 'D', 47, 6),
 (49, 'A', 48, 7),
 (50, 'A', 49, 7),
 (51, 'A', 50, 7),
 (52, 'A', 51, 7),
 (53, 'D', 52, 7),
 (54, 'A', 53, 8),
 (55, 'A', 54, 8),
 (56, 'A', 55, 8),
 (57, 'A', 56, 8),
 (58, 'A', 57, 8),
 (59, 'A', 58, 8),
 (60, 'D', 59, 8),
 (61, 'A', 60, 9),
 (62, 'A', 61, 9),
 (63, 'A', 62, 9),
 (64, 'A', 63, 9),
 (65, 'D', 60, 9),
 (66, 'A', 64, 10),
 (67, 'A', 65, 10),
 (68, 'A', 66, 10),
 (69, 'A', 67, 10),
 (70, 'A', 68, 10),
 (71, 'D', 69, 10),
 (72, 'A', 70, 11),
 (73, 'A', 71, 11),
 (74, 'A', 72, 11),
 (75, 'A', 73, 11),
 (76, 'A', 74, 11),
 (77, 'A', 75, 11),
 (78, 'A', 76, 11),
 (79, 'A', 77, 11),
 (80, 'A', 78, 11),
 (81, 'A', 79, 11),
 (82, 'A', 80, 11),
 (83, 'A', 81, 11),
 (84, 'D', 82, 11),
 (85, 'A', 53, 12),
 (86, 'A', 83, 12),
 (87, 'A', 84, 12),
 (88, 'A', 85, 12),
 (89, 'D', 86, 12),
 (90, 'A', 87, 13),
 (91, 'A', 88, 13),
 (92, 'A', 89, 13),
 (93, 'A', 90, 13),
 (94, 'A', 91, 13),
 (95, 'D', 92, 13),
 (96, 'A', 93, 14),
 (97, 'A', 94, 14),
 (98, 'A', 95, 14),
 (99, 'A', 96, 14),
 (100, 'A', 97, 14),
 (101, 'D', 98, 14),
 (102, 'A', 99, 15),
 (103, 'A', 100, 15),
 (104, 'A', 101, 15),
 (105, 'A', 102, 15),
 (106, 'A', 103, 15),
 (107, 'A', 104, 15),
 (108, 'D', 105, 15),
 (109, 'A', 106, 16),
 (110, 'A', 107, 16),
 (111, 'A', 108, 16),
 (112, 'A', 77, 16),
 (113, 'A', 109, 16),
 (114, 'A', 110, 16),
 (115, 'D', 106, 16),
 (116, 'D', 109, 16),
 (117, 'A', 111, 17),
 (118, 'A', 112, 17),
 (119, 'A', 113, 17),
 (120, 'D', 114, 17),
 (121, 'A', 115, 18),
 (122, 'A', 116, 18),
 (123, 'A', 117, 18),
 (124, 'A', 118, 18),
 (125, 'A', 119, 18),
 (126, 'A', 120, 18),
 (127, 'D', 121, 18),
 (128, 'A', 122, 19),
 (129, 'A', 123, 19),
 (130, 'A', 124, 19),
 (131, 'A', 125, 19),
 (132, 'D', 126, 19),
 (133, 'A', 127, 20),
 (134, 'A', 117, 20),
 (135, 'A', 128, 20),
 (136, 'A', 129, 20),
 (137, 'A', 130, 20),
 (138, 'A', 28, 20),
 (139, 'D', 131, 20),
 (140, 'A', 132, 21),
 (141, 'A', 100, 21),
 (142, 'A', 133, 21),
 (143, 'A', 134, 21),
 (144, 'A', 135, 21),
 (145, 'D', 136, 21),
 (146, 'A', 137, 22),
 (147, 'A', 138, 22),
 (148, 'A', 139, 22),
 (149, 'A', 140, 22),
 (150, 'A', 141, 22),
 (151, 'A', 142, 22),
 (152, 'A', 143, 22),
 (153, 'A', 144, 22),
 (154, 'A', 145, 22),
 (155, 'A', 146, 22),
 (156, 'D', 147, 22),
 (157, 'A', 148, 4),
 (158, 'A', 149, 23),
 (159, 'A', 150, 23),
 (160, 'A', 151, 23),
 (161, 'A', 152, 23),
 (162, 'D', 153, 23)]

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [29]:
%%sql
SELECT
    MTitle, TName, Showtime, CCode, CName, COUNT(*)
FROM
    DATASET
GROUP BY
    MTitle, TName, Showtime, CCode, CName
HAVING 
    COUNT(*) > 1

 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner', 2)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [30]:
%%sql
DROP TABLE DATASET

 * sqlite:///Movies.db
Done.


[]

### LET's check its gone

In [32]:
%%sql
SELECT *
    FROM DATASET;

 * sqlite:///Movies.db
(sqlite3.OperationalError) no such table: DATASET [SQL: 'SELECT *\n    FROM DATASET;'] (Background on this error at: http://sqlalche.me/e/e3q8)


### Yay it's GONE 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".